mpu6050/mpu6050_main.c

```diff
@@ -0,0 +1,373 @@
+			index = count - 1;
+		}
+	}
+	return index;
+}
+
+static void read_timer(unsigned long data);
+
+/* Module initialization */
+static struct kobject *g_kobject;
+static struct task_struct *g_reading_thread;
+DECLARE_COMPLETION(g_read_comletion);
+DEFINE_TIMER(g_read_timer, read_timer, 0, 0);
+DEFINE_SPINLOCK(g_read_timer_lock);
+#define TIMER_DELAY_MSEC 1000
+
+void read_timer(unsigned long param)
+{
+	complete(&g_read_comletion);
+	pr_info("mpu6050 %s: timer fired:\n", __func__);
+}
+
+int reading_thread(void *data)
+{
+	u64 time;
+	int ret;
+
+	pr_info("mpu6050 %s started\n", __func__);
+
+	while (!kthread_should_stop()) {
+		pr_info("mpu6050 %s waiting\n", __func__);
+		wait_for_completion(&g_read_comletion);
+		if (!kthread_should_stop())
+			mpu6050_read_data(false);
+		else
+			break;
+		time = get_jiffies_64() + msecs_to_jiffies(TIMER_DELAY_MSEC);
+		ret = mod_timer(&g_read_timer, time);
+		pr_info("mpu6050 %s: timer rescheduled: %d\n", __func__, ret);
+	}
+
+	g_reading_thread = NULL;
+	pr_info("mpu6050 %s finished\n", __func__);
+	return 0;
+}
+
+static void free_sysfs(void)
+{
+	if (g_kobject) {
+		sysfs_remove_files(g_kobject, g_attributes);
+		kobject_put(g_kobject);
+
+		pr_info("mpu6050: sysfs data destroyed\n");
+	}
+}
+
+static int __init mpu6050_init(void)
+{
+	int ret;
+	u64 timer_time;
+	struct cdevs_holder *cdevs = get_cdevs();
+
+	/* Create i2c driver */
+	ret = i2c_add_driver(&mpu6050_i2c_driver);
+	if (ret) {
+		pr_err("mpu6050: failed to add new i2c driver: %d\n", ret);
+		return ret;
+	}
+	pr_info("mpu6050: i2c driver created\n");
+
+	g_kobject = kobject_create_and_add("mpu6050", kernel_kobj);
+	if (!g_kobject) {
+		ret = -ENOMEM;
+		pr_err("mpu6050: failed to create sysfs kobject: %d\n", ret);
+		goto error1;
+	}
+
+	ret = sysfs_create_files(g_kobject, g_attributes);
+	if (ret) {
+		pr_err("mpu6050: failed to create sysfs data attributes: %d\n",
+			ret);
+		goto error2;
+	}
+	pr_info("mpu6050: sysfs data attributes created\n");
+
+	init_mpu6050_data(
+		get_mpu6050_data(), ELEMENTS_COUNT, mpu6050_read_data);
+
+	ret = init_cdevs(cdevs, get_mpu6050_data());
+	if (ret) {
+		pr_err("mpu6050: failed to create cdevs: %d\n", ret);
+		goto error3;
+	}
+
+	g_reading_thread = kthread_run(reading_thread, NULL, "mpu6050_read_th");
+	if (!g_reading_thread) {
+		pr_err("mpu6050: failed to create reading thread\n");
+		goto error4;
+	}
+
+	init_timer_deferrable(&g_read_timer);
+	timer_time = get_jiffies_64() + msecs_to_jiffies(TIMER_DELAY_MSEC);
+	ret = mod_timer(&g_read_timer, timer_time);
+	pr_info("mpu6050: timer init: %d\n", ret);
+
+	pr_info("mpu6050: module loaded\n");
+	return 0;
+
+error4:
+	free_cdevs(get_cdevs());
+error3:
+	free_mpu6050_data(get_mpu6050_data());
+error2:
+	free_sysfs();
+error1:
+	i2c_del_driver(&mpu6050_i2c_driver);
+	return ret;
+}
+
+static void __exit mpu6050_exit(void)
+{
+	int result;
+
+	if (g_reading_thread) {
+		complete(&g_read_comletion);
+		kthread_stop(g_reading_thread);
```

Comment:
It makes sense to set should_stop flag by `kthread_stop` before waking up the thread by `complete`.

Reply:
Done.